presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteSpatialPartitioningAggregate.java

```diff
@@ -0,0 +1,105 @@
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static java.lang.String.format;
+
+public class RewriteSpatialPartitioningAggregate
+        implements Rule<AggregationNode>
+{
+    private static final TypeSignature GEOMETRY_TYPE_SIGNATURE = parseTypeSignature("Geometry");
+    private static final String NAME = "spatial_partitioning";
+    private static final Signature NEW_SIGNATURE = new Signature(NAME, AGGREGATE, VARCHAR.getTypeSignature(), GEOMETRY_TYPE_SIGNATURE, DOUBLE.getTypeSignature(), INTEGER.getTypeSignature());
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(RewriteSpatialPartitioningAggregate::hasSpatialPartitioning);
+
+    private static boolean hasSpatialPartitioning(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations().values().stream()
+                .map(Aggregation::getCall)
+                .anyMatch(call -> call.getName().toString().equals(NAME) && call.getArguments().size() == 2);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode node, Captures captures, Context context)
+    {
+        ImmutableMap.Builder<Symbol, Aggregation> aggregations = ImmutableMap.builder();
+        Symbol partitionCountSymbol = context.getSymbolAllocator().newSymbol("partition_count", INTEGER);
+        for (Map.Entry<Symbol, Aggregation> entry : node.getAggregations().entrySet()) {
+            Aggregation aggregation = entry.getValue();
+            FunctionCall call = aggregation.getCall();
+            QualifiedName name = call.getName();
+            if (name.toString().equals(NAME) && call.getArguments().size() == 2) {
+                ImmutableList.Builder<Expression> arguments = ImmutableList.builder();
+                arguments.addAll(call.getArguments());
+                arguments.add(partitionCountSymbol.toSymbolReference());
+                aggregations.put(entry.getKey(), new Aggregation(new FunctionCall(name, arguments.build()), NEW_SIGNATURE, aggregation.getMask()));
+            }
+            else {
+                aggregations.put(entry);
+            }
+        }
+
+        ProjectNode projectNode = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                node.getSource(),
+                Assignments.builder()
+                        .putIdentities(node.getSource().getOutputSymbols())
+                        .put(partitionCountSymbol, new LongLiteral(format("%s", getHashPartitionCount(context.getSession()))))
+                        .build());
+
+        return Result.ofPlanNode(new AggregationNode(
```

Comment:
nit: put `new AggregationNode(` on the next line for better "plan tree" readability